railties/lib/rails/generators/rails/app/templates/config/puma.rb.tt

```diff
@@ -2,38 +2,53 @@
 # are invoked here are part of Puma's configuration DSL. For more information
 # about methods provided by the DSL, see https://puma.io/puma/Puma/DSL.html.
 
-# Puma can serve each request in a thread from an internal thread pool.
-# The `threads` method setting takes two numbers: a minimum and maximum.
-# Any libraries that use thread pools should be configured to match
-# the maximum value specified for Puma. Default is set to 5 threads for minimum
-# and maximum; this matches the default thread size of Active Record.
-max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
-min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
-threads min_threads_count, max_threads_count
-
-# Specifies that the worker count should equal the number of processors in production.
-if ENV["RAILS_ENV"] == "production"
+rails_env = ENV.fetch("RAILS_ENV", "development")
+
+# Puma starts a configurable number of processes (workers) and each process
+# serves each request in a thread from an internal thread pool.
+#
+# As a rule of thumb, increasing the number of threads will increase how much
+# traffic a given process can handle (throughput), but due to Ruby's
+# Global VM Lock (GVL) it has diminishing returns, and will degrade the
+# response time (latency) of the application.
+#
+# The ideal number of threads per worker depend both how much time the
+# application spend waiting for IO operations, and how much you wish to
+# to prioritize throughput over latency.
+#
+# The default is set to 3 threads as it's deemed a decent compromise between
+# throughput and latency for the average Rails application.
+#
+# Any libraries that use a connection pool or another resource pool should
+# be configured to provide at least as many connections as the number of
+# threads. This includes Active Record's `pool` parameter in `database.yml`.
+threads_count = ENV.fetch("RAILS_MAX_THREADS") { 3 }
+threads threads_count, threads_count
+
+if rails_env == "production"
+  # Specifies that the worker count should equal the number of processors in production.
+  #
+  # The number of processor is detected using various platform dependent methods, but
+  # note that this isn't very reliable, especially in containerized environments like Docker.
+  # As such, it is recommended to set it manually.
   require "concurrent-ruby"
   worker_count = Integer(ENV.fetch("WEB_CONCURRENCY") { Concurrent.physical_processor_count })
```

Comment:
@byroot Not strictly related to the original issue, but what do you think about making it `Concurrent.processor_count`? The value for both methods differ in environments such as AWS EC2, etc.

Reply:
Yeah, this is a thought one. 

So I just did a bit of spelunking, it was added in https://github.com/rails/rails/pull/46838, there is not much precision on why limiting to physical cores was chosen.

To be honest I kinda share many of the worries in that thread, lots of platform will report way more cores (physical or not) than the user can actually use, and AFAIK there really isn't a proper fix for that. `make -j` or similar running in a container ending up spawning way too many processes is pretty much a running gag nowadays...

Hence why I added the comment about how important it is to set it manually.

Now as to `processor_count` vs `physical_processor_count`, again, not sure about the original intent, but I can get behind it given that hyper threads are far from giving the same performance than physical cores, so depending on what your priorities are, it may make sense to ignore them. But I suppose it's one more of these things it's hard to argue about one being better than the other, hence hard to provide a default that will satisfy most users.

Reply:
Unfortunately, we live in a hypercore-driven world. When AWS sells you 2vCPU, they're selling you two logical cores and once physical.

I've seen enough benchmarks on cloud providers and using logical core count rather than physical to say I'm pretty confident that setting to the number of logical cores is probably the right choice. I don't have any ways to address concerns about the platform reporting silly numbers though 😬 

> hyper threads are far from giving the same performance than physical cores

We've been running 1 proc per hyperthread in the cloud for the last decade, it seems to work fine. IIRC this is how @noahgibbs ran rails-ruby-bench. But hey, maybe we can do another benchmark thread? 😆

Reply:
Come to think of it, I wonder if @dhh's original benchmark from the thread was reporting physical or logical cores...

Reply:
Yeah, thinking about it, I don't think it makes much sense to ignore logical cores, we might as well default to `Etc.nprocessors`. But would be nice to check it works properly at least on the most popular platforms like Heroku (I don't have an account so can't check).

There is a lot of controversy about hyperthread performance but ultimately it varies a lot from one chip to another so it's not like there is a default that make sense. If someone pays for 4vCPU they should run 4 processes most likely.

I'm even a bit tempted to consider this as a bug and backport it to 7-1-stable.

Reply:
`Etc.nprocessors` works fine on Heroku, but returns values you would never want to use (i.e. returns 8 for a Standard-1X dyno, and `Concurrent.physical_processor_count` returns 4, but on a Standard-1X you probably only want 1 or 2). For what it's worth, `Concurrent.processor_count` already delegates to `Etc.nprocessors` (see https://github.com/ruby-concurrency/concurrent-ruby/pull/814).

Reply:
> `Concurrent.processor_count` already delegates to `Etc.nprocessors` (see [ruby-concurrency/concurrent-ruby#814](https://github.com/ruby-concurrency/concurrent-ruby/pull/814)).

This is no longer true : https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/utility/processor_counter.rb#L34

Reply:
> `Etc.nprocessors` works fine on Heroku, but returns values you would never want to use (i.e. returns 8 for a Standard-1X dyno,

That's exactly what I meant by "not working", it's giving you the number of cores on the host, which is totally decorelated from the number of core the platform actually allow you to use.

Googling for it, it's even documented in their doc: https://devcenter.heroku.com/articles/deploying-rails-applications-with-the-puma-web-server#process-count-value

> The value for `nproc` from eco, basic, standard-1x, and standard-2x dynos are correct, but these cores are shared between multiple applications running in containers. While nproc for these dynos will all return 8, it is best to assume only one process can execute at a time.

That phrasing is amazing too... "It's correct, but don't use it because it's not what you asked for".

So yeah, we have a pretty big issue with that default. On bare metal it will ignore hyper-threads, so return half what you'd expect, and on many managed platforms it will return way too big a number.

As much as I like having the default auto-detected, given that this is pretty much an unsolved problem across the industry, I'll likely revert https://github.com/rails/rails/pull/46838

Reply:
Alright, I reverted it back to `1` as a default. Unfortunate, but better than spawning a silly amount of workers.